DOC: figure explanation (#24999)
matplotlib/matplotlib: doc/users/explain/figures.rst

`pyplot </tutorials/introductory/pyplot>` interface. As noted in
`api_interfaces`, the pyplot interface serves two purposes.  One is to spin
up the Backend and keep track of GUI windows. The other is a global state for
Axes and Artists that allow a short form API to plotting methods. In the
example above, we only use pyplot for the first purpose, to create the Figure
object, ``fig``.

In addition to `~.pyplot.Figure`, there are convenience methods to create Axes
at the same time that the figure is created since these two actions are so commonly
done at the same time. A simple grid can be achieved with `~.pyplot.subplots` (which
simply wraps `~.Figure.subplots`):

.. plot::
    :include-source:

    fig, axs = plt.subplots(2, 2, figsize=(4, 3), layout='constrained')

More complex grids can be achieved with `~.pyplot.subplot_mosaic` (which wraps
`~.Figure.subplot_mosaic`):

.. plot::
    :include-source:

    fig, axs = plt.subplot_mosaic([['A', 'right'], ['B', 'right']],
                                  figsize=(4, 3), layout='constrained')
    for ax_name in axs:
        axs[ax_name].text(0.5, 0.5, ax_name, ha='center', va='center')

> We can also have subfigures inside figures (or other subfigures;

Author: Added an example, which if we are going into why to use subfigures, seems the easiest way to explain it.

:doc:`/examples/subplots_axes_and_figures/subfigures` for more details).

It is possible to directly instantiate a `.Figure` instance.  This is usually
only necessary if you want to create your own GUI application and not use
the pyplot application.  See  the
embedding examples in :doc:`/gallery/user_interfaces/` for examples of
how to do this.

Figure options
--------------

There are a few options available when creating figures.  The Figure size on
the screen is set by *figsize* and *dpi*.  *figsize* is the ``(width, height)``
of the Figure in inches (or, if prefered, units of 72 typographic points).  *dpi*

> are how many pixels per inch the figure will be rendered at.  To make you Figures

Author: This was an explanation as to what *dpi* means.

> are how many pixels per inch the figure will be rendered at.  To make you Figures

Author: Yep caught that one I think

helpful to understand how to view a Figure.  This varies based on how you are
using Matplotlib, and what :ref:`Backend <what-is-a-backend>` you are using.

Notebooks and IDEs
------------------

If you are using a Notebook (e.g. `Jupyter <https://jupyter.org>`_) or an IDE
that renders Notebooks (PyCharm, VSCode, etc), then they have a backend that
will render the Matplotlib Figure when a code cell is executed.  One thing to
be aware of is that the default Jupyter backend (``%matplotlib inline``) will
by default trim or expand the figure size to have a tight box around Artists
added to the Figure (see :ref:`saving_figures`, below).

Standalone scripts and interctive use
-------------------------------------

If the user is on a client with a windowing system, there are a number of
:ref:`Backends <what-is-a-backend>` that can be used to render the Figure to
the screen, usually using a python Qt, Tk, or Wx toolkit, though there is a native
MacOS backend as well.  These are typically chosen either in the user's
:ref:`matplotlibrc <customizing-with-matplotlibrc-files>`, or by calling
``matplotlib.use('Qt5Agg')`` at the beginning of a session or script.

> When run from a script, or interactively (eg from an

Reviewer: ```suggestion
When run from a script, or interactively (e.g. from an
```
(although I am very open to that this OK nowadays)

above, we only use pyplot for the first purpose, to create the Figure object, ``fig``.

In addition to `.pyplot.Figure`, there are convenience methods to create Axes
at the same time that the figure is created since these two actions are so commonly
done at the same time. A simple grid can be achieved with `.pyplot.subplots` (which
simply wraps `.Figure.subplots`):

.. plot::
    :include-source:

    fig, axs = plt.subplots(2, 2, figsize=(4, 3), layout='constrained')

More complex grids can be achieved with `.pyplot.subplot_mosaic` (which wraps
`.Figure.subplot_mosaic`):

.. plot::
    :include-source:

    fig, axs = plt.subplot_mosaic([['A', 'right'], ['B', 'right']],
                                  figsize=(4, 3), layout='constrained')
    for ax_name in axs:
        axs[ax_name].text(0.5, 0.5, ax_name, ha='center', va='center')

We can also have subfigures inside figures (or other subfigures;
:ref:`/examples/subplots_axes_and_figures/subfigures` for more details).

It is possible to directly instantiate a `.Figure` instance.  This is usually
only necessary if you want to create your own GUI application and not use

> the pyplot application.  See  see the

Reviewer: ```suggestion
the pyplot application.  See the
```

how to do this.

Figure options
--------------

There are a few options available when creating figures.  The Figure size on
the screen is set by *figsize* and *dpi*.  *figsize* is the ``(width, height)``
of the Figure in inches (or, if prefered, units of 72 typographic points).  *dpi*
are how many pixels per inch the figure will be rendered at.  To make you Figures
appear on the screen at the physical size you requested, you should set *dpi*
to the same *dpi* as your graphics system.  Note that many graphics systems now use
a "dpi ratio" to specify how many screen pixels are used to represent a graphics
pixel.  Matplotlib applies the dpi ratio to the *dpi* passed to the figure to make
it have higher resolution, so you should pass the lower number to the figure.

The *facecolor*, *edgecolor*, *linewidth*, and *frameon* options all change the appearance of the
figure in expected ways, with *frameon* making the figure transparent if set to *False*.

Finally, the user can specify a layout engine for the figure with the *layout*
parameter.  Currently Matplotlib supplies
:ref:`"constrained" </tutorials/intermediate/constrainedlayout_guide>`,
:ref:`"compressed" <compressed_layout>` and

> `"tight" </tutorials/intermediate/tight_layout_guide>` layout engines.  These

Reviewer: ```suggestion
:ref:`"tight" </tutorials/intermediate/tight_layout_guide>` layout engines.  These
```
(although the link is probably not in correct format, not sure which is though...)

.. _figure_explanation:

================================================
Creating, viewing, and saving Matplotlib Figures
================================================

.. plot::
    :include-source:

    fig = plt.figure(figsize=(2, 2), facecolor='0.8', layout='constrained')
    ax = fig.add_subplot()

When looking at Matplotlib visualization, you are almost always looking at

> Artists placed on a `~.Figure`.  In the example above, the figure is the

Reviewer 2: xref to the "parts of a figure" document?

================

We will discuss how to create Figures in more detail below, but first it is
helpful to understand how to view a Figure.  This varies based on how you are
using Matplotlib, and what :ref:`Backend <what-is-a-backend>` you are using.

Notebooks and IDEs
------------------

If you are using a Notebook (e.g. `Jupyter <https://jupyter.org>`_) or an IDE
that renders Notebooks (PyCharm, VSCode, etc), then they have a backend that
will render the Matplotlib Figure when a code cell is executed.  One thing to
be aware of is that the default Jupyter backend (``%matplotlib inline``) will
by default trim or expand the figure size to have a tight box around Artists
added to the Figure (see :ref:`saving_figures`, below).

Standalone scripts and interctive use
-------------------------------------

If the user is on a client with a windowing system, there are a number of
:ref:`Backends <what-is-a-backend>` that can be used to render the Figure to
the screen, usually using a Python Qt, Tk, or Wx toolkit, though there is a native
MacOS backend as well.  These are typically chosen either in the user's
:ref:`matplotlibrc <customizing-with-matplotlibrc-files>`, or by calling

> ``matplotlib.use('Qt5Agg')`` at the beginning of a session or script.

Reviewer 2: ```suggestion
``matplotlib.use('QtAgg')`` at the beginning of a session or script.
```

When run from a script, or interactively (e.g. from an
`iPython shell <https://https://ipython.readthedocs.io/en/stable/>`_) the Figure

> will not be shown until we call ``plt.show()``. The Figure will appear in

Reviewer 2: this should xref to the `plt.ion` explanation.

a new GUI window, and usually will have a toolbar with Zoom, Pan, and other tools
for interacting with the Figure.  By default, ``plt.show()`` blocks
further interaction from the script or shell until the Figure window is closed,
though that can be toggled off for some pusposes.

Note that if you are on a client that does not have access to a windowing
system, the Figure will fallback to being drawn using the "Agg" backend, and
cannot be viewed, though it can be :ref:`saved <saving_figures>`.

.. _creating_figures:

Creating Figures
================

By far the most common way to create a figure is using the
:doc:`pyplot </tutorials/introductory/pyplot>` interface. As noted in
:ref:`api_interfaces`, the pyplot interface serves two purposes.  One is to spin
up the Backend and keep track of GUI windows. The other is a global state for

> Axes and Artists that allow a short form API to plotting methods. In the

Author: It can be opted out of when you don't use pyplot, can't it?  I'm not quite following your concern here.

More complex grids can be achieved with `~.pyplot.subplot_mosaic` (which wraps
`~.Figure.subplot_mosaic`):

.. plot::
    :include-source:

    fig, axs = plt.subplot_mosaic([['A', 'right'], ['B', 'right']],
                                  figsize=(4, 3), layout='constrained')
    for ax_name in axs:
        axs[ax_name].text(0.5, 0.5, ax_name, ha='center', va='center')

Sometimes we want to have a nested layout in a Figure, with two or more sets of
Axes that do not necessarily share the same subplot grid.
We can use `~.Figure.add_subfigure` or `~.Figure.subfigures` to create virtual
figures inside a parent Figure; see
:doc:`/gallery/subplots_axes_and_figures/subfigures` for more details.

.. plot::
    :include-source:

    fig = plt.figure(layout='constrained', facecolor='beige')
    fig.suptitle('Figure')
    figL, figR = fig.subfigures(1, 2)
    figL.set_facecolor('0.9')
    axL = figL.subplots(2, 1, sharex=True)
    figL.suptitle('Left subfigure')
    figR.set_facecolor('0.7')
    axR = figR.subplots(1, 2, sharey=True)
    figR.suptitle('Right subfigure')

It is possible to directly instantiate a `.Figure` instance.  This is usually

> only necessary if you want to create your own GUI application and not use

Reviewer 2: I think we should include a note that this is how to opt-out of the global registry of `Figure` s that `pyplot` does and is a good choice for batch or webserver applications where a given process is expected to create many `Figure` objects through its life.

Author: I didn't want to get into a lot of use cases here - presumably folks who need this are sophisticated enough to not need this document at all.  Here is what I changed it to:

```
This is usually only necessary if you want to create your
own GUI application or service that you do not want carrying the pyplot global
state.
```

.. _figure_explanation:

================================================
Creating, viewing, and saving Matplotlib Figures
================================================

> 

Reviewer 3: We should define what a figure is. Maybe something like

> In matplotlib, a `~.Figure` is the logical container for all visual elements. Think of it like a blank sheet of paper.
> To add content, you'll add visual elements (called `.Artist`\s) to the figure though specific functions.

This is somewhat informal, but should give a rough mental model.

Author: I think thats what the first two sentences below the example were meant to do?

Reviewer 3: > When looking at Matplotlib visualization, you are almost always looking at
Artists placed on a `~.Figure`. 

Has too little emphasis on *Figure*. *Matplotlib visualization* and *Artists* are mentioned first and are the syntactic objects in the sentence. *Figure* is only mentioned as the context of *Artists*. I would like a sentence that has *Figure* as the subject, like "A Figure is...". Also, I think this should come before any plot to set the scope/expectation. The plot code already has a lot of details and the user does not know what the relevance of the plot/shown code is.

Author: As below, readers are unlikely to hit this page with no context.  They will have seen matplotlib visualizations, even if only on the landing page.  And the title clearly indicates that this is about Figures.   We need to put a figure in context of what they have seen before so I prefer to mention that first.

.. plot::
    :include-source:

    fig = plt.figure(figsize=(2, 2), facecolor='0.8', layout='constrained')

>     ax = fig.add_subplot()

Reviewer 3: Maybe add first
```
    fig = plt.figure(figsize=(2, 2), facecolor='0.8')
    fig.text(0.1, 0.5, "This is a figure")
```
to illustrate the basic conecpt of adding content with out the need to already dive into the `Axes` / subplot complexity.

And then follow up with the Axes discussion.

Author: I'm not in favour of adding a relatively obscure artist to the first example - an Axes is what 99.99% of visualizations start with.  I personally wasn't even aware that Figure has a text method, so it's pretty obscure.

Author: Though this inspires a new section below on adding artists....

Reviewer 3: I'm trying to make a minimal illustration of a figure. Maybe even
```
plt.figure(facecolor='lightblue')
```
is enough.

Side-remark: Please don't use the somewhat obscure '0.8' greyscale notation in a basic example. That is unnecessarily confusing.

If the empty figure is too little, you could use `suptitle()`, which is simpler than `text()`:

```
plt.figure(facecolor='lightblue')
plt.suptitle('This is a figure')
```

Author: I want a minimal typical example of a figure in the context of what the reader has likely seen before, not an artificial example with nothing on it.  Readers are not starting with zero context here.  

I can easily change the background colour.

Author: Updated to change the color and added some annotation on the Axes.

Reviewer 3: > I want a minimal typical example of a figure in the context of what the reader has likely seen before, not an artificial example with nothing on it.

I would agree for a tutorial. In explanations however, I think we should clearly isolate the figure. Both visually, and from the code complexity

```
fig = plt.figure(facecolor='lightblue')
fig.suptitle('This is a figure')
```
![grafik](https://user-images.githubusercontent.com/2836374/213865657-c9cc31d0-7888-4e65-b309-a62dbf637905.png)

is much clearer than

```
fig = plt.figure(figsize=(2, 2), facecolor='lightskyblue',
                 layout='constrained')
fig.suptitle('Figure')
ax = fig.add_subplot()
ax.set_title('Axes', loc='left', fontstyle='oblique', fontsize='medium')
```

![grafik](https://user-images.githubusercontent.com/2836374/213865603-5a54c563-e3a5-4783-838f-c338589141e3.png)

> Readers are not starting with zero context here.

Author: Again, I think this initial example is more effective in the context of something the viewer would have seen - 99.9999% of figures have an Axes.  The complexity is all in separating the two objects in code so that the reader can understand how Matplotlib differentiates them internally.  I don't feel that an empty figure is a good use of space or the reader's time.

Reviewer 3: I beg to differ on the overall statement. 

> in the context of something the viewer would have seen - 99.9999% of figures have an Axes.

Note also that `plt.figure(); fig.add_subplot()` is rarely used (`plt.subplots()` being far more common and IMHO recommended). There is value in using the figure() call here explicitly and I'm not suggesting to change, but this fact weakens the familiarity argument. The generated figure is familiar, but the associated code is not.

I have the feeling that the current approach is neither fish nor fowl. Maybe, if you wan't to start with the familiar, really do subplots() and then explain that this does the two statements under the hood.

I really appreciate your work on the docs, and I'd like to make sure they are of maximal use for a diverse audience. I would like to encourage  to carefully consider how the descriptions work for users with different backgrounds. We as authors have a very good understanding, but throwing in variations (add_subplot vs subplots) or concepts (In the text Artists are mentioned without explanation or even only a link for that) can be obstacles for some readers.

Unfortunately, I don't have much time currently. So I will withdraw from the PR and the discussion.

``matplotlib.use('QtAgg')`` at the beginning of a session or script.

When run from a script, or interactively (e.g. from an
`iPython shell <https://https://ipython.readthedocs.io/en/stable/>`_) the Figure
will not be shown until we call ``plt.show()``. The Figure will appear in
a new GUI window, and usually will have a toolbar with Zoom, Pan, and other tools
for interacting with the Figure.  By default, ``plt.show()`` blocks
further interaction from the script or shell until the Figure window is closed,
though that can be toggled off for some pusposes.  For more details, please see
:ref:`controlling-interactive`.

Note that if you are on a client that does not have access to a windowing
system, the Figure will fallback to being drawn using the "Agg" backend, and
cannot be viewed, though it can be :ref:`saved <saving_figures>`.

.. _creating_figures:

Creating Figures
================

By far the most common way to create a figure is using the
:doc:`pyplot </tutorials/introductory/pyplot>` interface. As noted in
:ref:`api_interfaces`, the pyplot interface serves two purposes.  One is to spin
up the Backend and keep track of GUI windows. The other is a global state for

> Axes and Artists that allow a short-form API to plotting methods. In the
> example above, we only use pyplot for the first purpose, to create the Figure
> object, ``fig``.

Reviewer 2: Continuing the discussion from https://github.com/matplotlib/matplotlib/pull/24999#discussion_r1072812820

> we only use pyplot for the first purpose, ...

That reads to me like "In the above example we are not adding the figures to the global registry".  Maybe something like:

```suggestion
Axes and Artists that allow a short-form API to plotting methods. In the
example above, we only want the first purpose, to create the Figure
object, ``fig``, but it is still added to the global registry to support the
second.
```

Author: OK, I understand now.  If we digress like this, I made it a bit more wordy:

```
By far the most common way to create a figure is using the
:doc:`pyplot </tutorials/introductory/pyplot>` interface. As noted in
:ref:`api_interfaces`, the pyplot interface serves two purposes.  One is to spin
up the Backend and keep track of GUI windows. The other is a global state for
Axes and Artists that allow a short-form API to plotting methods. In the
example above, we use pyplot for the first purpose, and create the Figure object,
``fig``. As a side effect ``fig`` is also added to pyplot's global state, and
can be accessed via `~.pyplot.gcf`.
```

:doc:`/tutorials/advanced/transforms_tutorial` for more details).

.. _saving_figures:

Saving Figures
==============

Finally, Figures can be saved to disk using the `~.Figure.savefig` method.
``fig.savefig('MyFigure.png', dpi=200)`` will save a PNG formatted figure to
the file ``MyFigure.png`` in the current directory on disk with 200 dots-per-inch
resolution.  Note that the filename can include a relative or absolute path to
any place on the file system.

Many types of output are supported, including raster formats like PNG, GIF, JPEG,
TIFF and vector formats like PDF, EPS, and SVG.

By default, the size of the saved Figure is set by the Figure size (in inches) and, for the raster
formats, the *dpi*.  If *dpi* is not set, then the *dpi* of the Figure is used.
Note that *dpi* still has meaning for vector formats like PDF if the Figure includes
Artists that have been :doc:`rasterized </gallery/misc/rasterization_demo>`; the
*dpi* specified will be the resolution of the rasterized objects.

It is possible to change the size of the Figure using the *bbox_inches* argument
to savefig.  This can be specified manually, again in inches.  However, by far

> the most common use is ``bbox_inches='tight'``.  This option trims, or expands, the
> size of the figure so that it is tight around all the artists in a figure, with a

Reviewer 2: Super picky word choice preference, feel free to disregard:

```suggestion
the most common use is ``bbox_inches='tight'``.  This option "shrink wraps,
trimming or expanding as needed, the size of the figure so that it is tight 
around all the artists in a figure, with a
```

Artists placed on a `~.Figure`.  In the example above, the figure is the
blue region and `~.Figure.add_subplot` has added an `~.axes.Axes` artist to the
`~.Figure` (see :ref:`figure_parts`).  A more complicated visualization can add
multiple Axes to the Figure, colorbars, legends, annotations, and the Axes
themselves can have multiple Artists added to them
(e.g. ``ax.plot`` or ``ax.imshow``).

.. _viewing_figures:

Viewing Figures
================

We will discuss how to create Figures in more detail below, but first it is
helpful to understand how to view a Figure.  This varies based on how you are
using Matplotlib, and what :ref:`Backend <what-is-a-backend>` you are using.

Notebooks and IDEs
------------------

If you are using a Notebook (e.g. `Jupyter <https://jupyter.org>`_) or an IDE
that renders Notebooks (PyCharm, VSCode, etc), then they have a backend that
will render the Matplotlib Figure when a code cell is executed.  One thing to
be aware of is that the default Jupyter backend (``%matplotlib inline``) will
by default trim or expand the figure size to have a tight box around Artists
added to the Figure (see :ref:`saving_figures`, below).

> Standalone scripts and interctive use

Reviewer 4: ```suggestion
Standalone scripts and interactive use
```

-------------------------------------

If the user is on a client with a windowing system, there are a number of
:ref:`Backends <what-is-a-backend>` that can be used to render the Figure to
the screen, usually using a Python Qt, Tk, or Wx toolkit, though there is a native
MacOS backend as well.  These are typically chosen either in the user's
:ref:`matplotlibrc <customizing-with-matplotlibrc-files>`, or by calling
``matplotlib.use('QtAgg')`` at the beginning of a session or script.

When run from a script, or interactively (e.g. from an
`iPython shell <https://https://ipython.readthedocs.io/en/stable/>`_) the Figure
will not be shown until we call ``plt.show()``. The Figure will appear in
a new GUI window, and usually will have a toolbar with Zoom, Pan, and other tools
for interacting with the Figure.  By default, ``plt.show()`` blocks
further interaction from the script or shell until the Figure window is closed,

> though that can be toggled off for some pusposes.  For more details, please see

Reviewer 4: ```suggestion
though that can be toggled off for some purposes.  For more details, please see
```

Author: Thanks, for some reason I had my spellchecker off, but now its on again ;-)
